### PR TITLE
[Autobuild] improve robustness and reduce cost

### PR DESCRIPTION
**Pull request by LeoLjl**
## Why are these changes needed?
In this PR, we introduce changes to agent builder. The changes are necessary for merging MetaAgent in future PR. The following changes are made:

1. Refine the prompts for agent selection, agent generation, generated agent's system message.
2. Include the context of group chat in the prompt of agents in nested chat.
3. Update build_from_library method so that for a given building task, 
  (a) retrieve from the agent library based on semantic similarity between agent description and the task
  (b) prompt an llm to double check if the retrieved agent suits the task
  (c) if the search result does not fit the task, a new agent with its name and descriptions will be generated
4. Update the autobuild_basic.ipynb accordingly.

## Related issue number
This is a preliminary update for merging MetaAgent PR #1749 

## Checks

- [ ] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

**Comment by sonichi** [comment]
AutoBuild test error: https://github.com/microsoft/autogen/actions/runs/9452298641/job/26035171981?pr=2907

**Comment by krishnashed** [comment]
I had a logical question with Autobuild, it is not necessary that we always need 5 agents to be created to solve the given task right? It might just add up to the complexity of agent system. While the same task can be easily solved with lesser number of agents.

**Comment by LinxinS97** [comment]
> I had a logical question with Autobuild, it is not necessary that we always need 5 agents to be created to solve the given task right? It might just add up to the complexity of agent system. While the same task can be easily solved with lesser number of agents.

You can control the number of agents it creates through the `max_agents` when initializing AutoBuild.

**Comment by krishnashed** [comment]
> > I had a logical question with Autobuild, it is not necessary that we always need 5 agents to be created to solve the given task right? It might just add up to the complexity of agent system. While the same task can be easily solved with lesser number of agents.
> 
> You can control the number of agents it creates through the `max_agents` when initializing AutoBuild.

Ya, my question was more like, let max_agents be the upper limit of agents being created, but let the AI figure out if the job can be done in lesser number of agents than the max_agents. Do you think this feature adds value ?

**Comment by LinxinS97** [comment]
> > > I had a logical question with Autobuild, it is not necessary that we always need 5 agents to be created to solve the given task right? It might just add up to the complexity of agent system. While the same task can be easily solved with lesser number of agents.
> > 
> > 
> > You can control the number of agents it creates through the `max_agents` when initializing AutoBuild.
> 
> Ya, my question was more like, let max_agents be the upper limit of agents being created, but let the AI figure out if the job can be done in lesser number of agents than the max_agents. Do you think this feature adds value ?

Currently, our prompt is "Suggest no more than {max_agent} experts...". Any chance to improve this?

**Review by krishnashed — CHANGES_REQUESTED** [review]

File: autogen/agentchat/contrib/agent_builder.py [hunk 1]
```diff
@@ -463,18 +480,20 @@ def build(
                 "code_execution_config": code_execution_config,
             }
         )
-
-        return self._build_agents(use_oai_assistant, **kwargs)
+        _config_check(self.cached_configs)
+        return self._build_agents(use_oai_assistant, user_proxy=user_proxy, **kwargs)
 
     def build_from_library(
```
[the remark this answers is not in the document]
Comment (krishnashed) [reply autogen/agentchat/contrib/agent_builder.py:486]: the latest commit fixes the issue, resolving the conversation.

**Review by krishnashed — APPROVED** [review]
the latest commit fixes the issue, approving the changes.

**Comment by krishnashed** [comment]
Now we can merge this PR, lgtm.

**Comment by LinxinS97** [comment]
@sonichi Could you check and run the openai test?

**Comment by Deykeepcomn4me** [comment]
This is hackers hacking me  please contact fbi

**Pull request merged**